tty/zynq7000-uart/zynq7000-uart.c

```diff
@@ -0,0 +1,498 @@
+	}
+
+	return platformctl(&ctl);
+}
+
+
+static int uart_initAmbaClk(unsigned int dev)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_ambaclock;
+
+	ctl.ambaclock.dev = dev;
+	ctl.ambaclock.state = 1;
+
+	return platformctl(&ctl);
+}
+
+
+static int uart_initClk(void)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+
+	/* Set IO PLL as source clock and set divider:
+	 * IO_PLL / 0x14 :  1000 MHz / 20 = 50 MHz     */
+	ctl.devclock.dev = pctl_ctrl_uart_clk;
+	ctl.devclock.clkact0 = 0x1;
+	ctl.devclock.clkact1 = 0x1;
+	ctl.devclock.srcsel = 0;
+	ctl.devclock.divisor0 = 0x14;
+
+	return platformctl(&ctl);
+}
+
+
+static void uart_destroy(unsigned int n)
+{
+	uart_t *uart = &uart_common.uarts[n];
+
+	munmap((void *)uart->base, _PAGE_SIZE);
+	libtty_close(&uart->tty);
+	libtty_destroy(&uart->tty);
+	resourceDestroy(uart->cond);
+	resourceDestroy(uart->lock);
+}
+
+
+static int uart_init(unsigned int n, speed_t baud)
+{
+	libtty_callbacks_t callbacks;
+	uart_t *uart = &uart_common.uarts[n];
+
+	if (uart_setPin(info[n].rxPin) < 0 || uart_setPin(info[n].txPin) < 0 || uart_initAmbaClk(info[n].clk) < 0)
+		return -EINVAL;
+
+	uart->base = mmap(NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE, OID_PHYSMEM, info[n].base);
+	if (uart->base == MAP_FAILED)
+		return -ENOMEM;
+
+	callbacks.arg = uart;
+	callbacks.set_cflag = uart_setCFlag;
+	callbacks.set_baudrate = uart_setBaudrate;
+	callbacks.signal_txready = uart_signalTXReady;
+
+	if (libtty_init(&uart->tty, &callbacks, _PAGE_SIZE) < 0) {
+		munmap((void *)uart->base, _PAGE_SIZE);
+		return -ENOENT;
+	}
+
+	if (condCreate(&uart->cond) != EOK) {
+		uart_destroy(n);
+		return -ENOENT;
+	}
+
+	if (mutexCreate(&uart->lock) != EOK) {
+		uart_destroy(n);
+		return -ENOENT;
+	}
+
+	/* Reset RX & TX */
+	*(uart->base + cr) = 0x3;
+
+	/* Disable TX and RX */
+	*(uart->base + cr) = (*(uart->base + cr) & ~0x000001ff) | 0x00000028;
+
+	uart_setBaudrate(uart, baud);
+	uart->tty.term.c_ispeed = uart->tty.term.c_ospeed = baud;
+
+	/* normal mode, 1 stop bit, no parity, 8 bits */
+	uart_setCFlag(uart, &uart->tty.term.c_cflag);
+
+	/* Uart Control Register
+	 * TXEN = 0x1; RXEN = 0x1; TXRES = 0x1; RXRES = 0x1 */
+	*(uart->base + cr) = (*(uart->base + cr) & ~0x000001ff) | 0x00000017;
+
+	/* Enable RX FIFO trigger */
+	*(uart->base + ier) |= 0x1;
+	/* Set trigger level, range: 1-63 */
+	*(uart->base + rxwm) = 1;
+
+	beginthread(uart_intThread, 4, &uart->stack, sizeof(uart->stack), (void *)uart);
+	interrupt(info[n].irq, uart_interrupt, uart, uart->cond, &uart->inth);
+
+	return EOK;
+}
+
+
+int main(int argc, char **argv)
+{
+	uart_t *uart;
+	speed_t baud = B115200;
+	uint32_t port;
+	int n, c, console = 1;
+
+	/* Get console ID and baudrate */
+	while ((c = getopt(argc, argv, "b:c:")) != -1) {
+		switch (c) {
+			case 'b':
+				if ((baud = libtty_int_to_baudrate(atoi(optarg))) == (speed_t)-1) {
+					debug("uart: wrong baudrate value\n");
+					return -1;
```

Comment:
As it turns out, we shouldn't return negative values from main
```suggestion
					return 1;
```

Reply:
Why? I have seen in our code that it is a common practice, some instances:
https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/fde0606cd220251d0c2572b34775d98796feea83/tty/uart16550/uart16550.c#L327-L330

https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/fde0606cd220251d0c2572b34775d98796feea83/multi/imxrt-multi/imxrt-multi.c#L430-L433

Reply:
Yeah, all of these are errors to be fixed.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/_exit.html
"The value of status may be 0, EXIT_SUCCESS, EXIT_FAILURE, [CX] [Option Start]  or any other value, though only the least significant 8 bits (that is, status & 0377) shall be available from wait() and waitpid(); (..)"
Only values from 0 to 255 are valid, rest of the bits are for OS purposes and are not preserved (so there is actually no way to successfully return negative value.